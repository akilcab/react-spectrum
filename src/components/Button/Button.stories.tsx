import { Story, Meta } from "@storybook/react/types-6-0";
import { Button, ButtonProps } from "./Button";
import { Omit } from "@material-ui/types";
import { ButtonProps as BtnProps } from "@material-ui/core/Button/Button";

export default {
  title: "Button",
  component: Button,
  description: `A button.`,
  argTypes: {
    background: { control: "color" },
    color: { control: "color" },
    borderRadius: { control: "string" },
  },
} as Meta;

//👇 We create a “template” of how args map to rendering
const Template: Story<ButtonProps & Omit<BtnProps, keyof ButtonProps>> = (
  args
) => <Button {...args}>Click me</Button>;

//👇 Each story then reuses that template
export const Default = Template.bind({});
Default.args = {
  background: "none",
  border: "1px solid #13938F",
  color: "#0d4e4c",
};
