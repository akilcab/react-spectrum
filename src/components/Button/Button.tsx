import React from "react";
import { Omit } from "@material-ui/types";
import { Button as Btn, ButtonProps as BtnProps } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      background: (props: ButtonProps) =>
        props.background
          ? props.background
          : `linear-gradient(to left, ${theme.palette.secondary.main} 0%, ${theme.palette.primary.main} 100%)`,
      border: (props: ButtonProps) => (props.border ? props.border : "none"),
      borderRadius: (props: ButtonProps) =>
        props.borderRadius ? props.borderRadius : "4px",
      color: (props: ButtonProps) => (props.color ? props.color : "#fff"),
      height: (props: ButtonProps) => (props.height ? props.height : "initial"),
      width: (props: ButtonProps) => (props.width ? props.width : "initial"),
      margin: (props: ButtonProps) => (props.margin ? props.margin : "initial"),
      padding: (props: ButtonProps) =>
        props.padding ? props.padding : "0.438rem 2.438rem",
      textTransform: (props: ButtonProps) =>
        props.textTransform ? props.textTransform : "none",
      minWidth: (props: ButtonProps) => (props.minWidth ? props.minWidth : 0),
    },
  })
);

export interface ButtonProps {
  color?: string;
  background?: string;
  borderRadius?: number | string;
  border?: number | string;
  height?: number | string;
  width?: number | string;
  padding?: number | string;
  margin?: number | string;
  textTransform?: "lowercase" | "capitalize" | "none" | "uppercase";
  children: React.ReactNode;
  minWidth?: string | number;
}

export const Button: React.FC<
  ButtonProps & Omit<BtnProps, keyof ButtonProps>
> = (props) => {
  const { color, border, children, ...other } = props;
  const classes = useStyles(props);

  delete other.borderRadius;
  delete other.minWidth;

  return (
    <Btn className={classes.root} {...other}>
      {children}
    </Btn>
  );
};
