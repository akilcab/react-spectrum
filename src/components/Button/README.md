# Description of component

CustomButton communicate actions that users can take. They are typically placed throughout your UI, in places like:

* Dialogs
* Modal windows
* Forms
* Cards
* Toolbars

#### Example :

![](assets/desc.png)

## How to import

```node
import { CustomButton } from "app/shared/components/CustomButton";
```

## How to use

```node
....
    <CustomButton>Valider</CustomButton>
....
```

## Props

| Props       | Required | Description | Type |
| :---        | :---     | :---        |:---  |
| `children` | true | The content of the button. | `node` |
| `color` | false | Allows you to define the foreground color of a text element and its possible decorations. | `string` |
| `background` | false | Allows you to define the different values of the properties related to the management of the backgrounds of an element (color, image, origin, size, repetition, etc. | `string` |
| `borderRadius` | false | Allows you to define rounded corners for the border of an element. | `string` or `number` |
| `border` | false | Allows you to define the properties related to the border. | `string` or `number` |
| `height` | false | Allows you to set the height of the content box of an element. | `string` or `number` |
| `width` | false | Allows you to define the width of the content box of an element. | `string` or `number` |
| `padding` | false | Allows you to define the different filling distances on the four sides of an element. | `string` or `number` |
| `margin` | false | Sets the size of the margins on all four sides of the element. | `string` or `number` |
| `textTransform` | false | Defines how to use capital letters for the text of an element. | `'lowercase'` or `'capitalize'` or `'none'` or `'uppercase'` |
| `minWidth` | false | The min-width property is used to define the minimum width of a given element. | `string` or `number` |

## CSS properties we can override

You can override all CSS properties of the ButtonProps type of Material UI.
