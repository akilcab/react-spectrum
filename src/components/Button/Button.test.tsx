import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Button } from "./Button";

describe("<Button />", () => {
  test("rend correctement", () => {
    const component = renderer.create(<Button>Valider</Button>).toJSON();
    expect(component).toMatchSnapshot();
  });

  test("should have content texts in the component", () => {
    const component = render(<Button>VALIDER</Button>);
    expect(component.container).toHaveTextContent("VALIDER");
    expect(component.container).not.toHaveTextContent("NO_VALIDER");
  });

  test("try to display the profile without its require props", () => {
    expect(Button).toThrowError();
  });
});
