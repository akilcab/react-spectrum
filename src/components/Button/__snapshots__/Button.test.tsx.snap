// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rend correctement 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-root-1 makeStyles-root-2"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    Valider
  </span>
</button>
`;
